WorkQueue should inherit from SerialFunctionDispatcher

jyavenard opened the pull request:
#### 4cc06b4be14a60c0b72098fa3e32ccb98657d525
<pre>
WorkQueue should inherit from SerialFunctionDispatcher
<a href="https://bugs.webkit.org/show_bug.cgi?id=260779">https://bugs.webkit.org/show_bug.cgi?id=260779</a>
rdar://114535523

Reviewed by Kimmo Kinnunen.

Make WorkQueue inherit from SerialFunctionDispatcher so that it can be
used similarly to RunLoop and WorkerOrWorkletThread.

Needed for this change to compile:
- Export the whole class to get the multiple inheritance to
  work across binaries.
- When exporting WorkQueue, it will export function with OSObjectPtr&lt;dispatch_queue_t&gt;.
  This has different type in Obj-C and in C++. If the type is used in both,
  then there will be unresolved symbol since C++ type is used when creating
  the library (WorkQueue.cpp), as such remove method prototype from WorkQueue
  and have a single WorkQueue constructor taking an enum instead.

Ideally, we would implement only bug 261080, but there are futher issues
getting it done. So to not block other works depending on this change,
we push this version first.

* Source/WTF/wtf/FunctionDispatcher.h:
* Source/WTF/wtf/SuspendableWorkQueue.cpp:
(WTF::SuspendableWorkQueue::stateString):
* Source/WTF/wtf/SuspendableWorkQueue.h:
* Source/WTF/wtf/WorkQueue.cpp:
(WTF::WorkQueue::main):
(WTF::WorkQueue::dispatch):
(WTF::WorkQueue::assertIsCurrent const):
(WTF::ConcurrentWorkQueue::dispatch):
* Source/WTF/wtf/WorkQueue.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY): Deleted.
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueue::WorkQueue):
(WTF::WorkQueue::constructMainWorkQueue): Deleted.
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueue::WorkQueue):
(WTF::WorkQueue::constructMainWorkQueue): Deleted.

Canonical link: <a href="https://commits.webkit.org/267623@main">https://commits.webkit.org/267623@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b90fdcf937a5d7cac2dba4975d7970d3c6a4104c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17176 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18963 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16074 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17644 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18270 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17379 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17724 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19780 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14964 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22292 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14836 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15963 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15772 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20110 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16398 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16363 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13882 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18740 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15517 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4353 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4108 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19884 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19965 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16194 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4208 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
```diff
@@ -57,7 +57,7 @@ IOChannel::~IOChannel()
 void IOChannel::read(size_t offset, size_t size, WTF::WorkQueueBase& queue, Function<void(Data&, int error)>&& completionHandler)
 {
-    queue.dispatch([this, protectedThis = Ref { *this }, offset, size, completionHandler = WTFMove(completionHandler)] {
+    queue.dispatchInternal([this, protectedThis = Ref { *this }, offset, size, completionHandler = WTFMove(completionHandler)] {
```
These look very dubious. I think these could be tried to be fixed by using the correct type, e.g. WorkQueue?

jyavenard replied:
Unfortunately, not the way the method is used, it is passed both a ConcurrentWorkQueue and a WorkQueue. 
Only way to fix this issue would be to make either IOChannel::read/write a templated method, or (preferred as more elegant, logical) make WorkQueue inherit from ConcurrentWorkQueue.
I've put such implementation here https://github.com/WebKit/WebKit/compare/main...jyavenard:WebKit:WorkQueue
to be discussed further

kkinnunen-apple commented on Source/WTF/wtf/SuspendableWorkQueue.h:
```diff
@@ -40,14 +40,22 @@ class SuspendableWorkQueue final : public WorkQueue {
     WTF_EXPORT_PRIVATE void suspend(Function<void()>&& suspendFunction, CompletionHandler<void()>&& suspensionCompletionHandler);
     WTF_EXPORT_PRIVATE void resume();
-    WTF_EXPORT_PRIVATE void dispatch(Function<void()>&&) final;
+    void dispatch(Function<void()>&& function) final
```
I'm not sure these virtual functions should be defined inline.
The issue is that libWTF.a has to have an definition, but then the use-sites in WebCore.framework might also generate a definition. Seems a bit too risky for what it presumably tries to achieve?

jyavenard replied:
I made a note on why it's done that way in the commit log, and discussion is on slack. Issue is with a JavaScriptCore utilities reading symbols, it gives error when that particular method isn't inline. no idea why.

jyavenard replied:
FWIW, it seems to be a bug in the `tapi` utility

kkinnunen-apple commented on Source/WTF/wtf/WorkQueue.h:
```diff
@@ -41,13 +41,13 @@
+    WTF_EXPORT_PRIVATE virtual ~WorkQueueBase();
 
-    WTF_EXPORT_PRIVATE void dispatch(Function<void()>&&) override;
+    WTF_EXPORT_PRIVATE void dispatchInternal(Function<void()>&&);
```
It doesn't sound like dispatchInternal, internal function, should be public.

I think we typically call this dispatchImpl().

kkinnunen-apple commented on Source/WTF/wtf/WorkQueue.h:
```diff
@@ -86,23 +86,25 @@ class WorkQueueBase : public FunctionDispatcher, public ThreadSafeRefCounted<Wor
  * That is, two different runnables dispatched to the WorkQueue should never be allowed to execute simultaneously.
  * They may be executed on different threads but can safely be used by objects that aren't already threadsafe.
- * Use `assertIsCurrent(m_myQueue);` in a runnable to assert that the runnable runs in a specific queue.
+ * Use `m_myQueue->assertIsCurrent();` in a runnable to assert that the runnable runs in a specific queue.
```
No, this is not the design.
The design is:
 - if something can be asserted as being current, it responds to assertIsCurrent(something). That something can be anything: thread, threadlikeassertion, runloop, workqueue, etc. So all have the same uniform interface.

jyavenard replied:
assertIsCurrent(m_myQueue) will continue to work just the same now that WorkQueue is a SerialFunctionDispatcher ; the assertIsCurrent implementation is in FunctionDispatcher.h

jyavenard replied:
I'll update the comment then.
But all supports `something->assertIsCurrent()` I believe

jyavenard committed:
WorkQueue should inherit from SerialFunctionDispatcher
https://bugs.webkit.org/show_bug.cgi?id=260779
rdar://114535523

Reviewed by Kimmo Kinnunen.

Make WorkQueue inherit from SerialFunctionDispatcher so that it can be
used similarly to RunLoop and WorkerOrWorkletThread.

Needed for this change to compile:
- Export the whole class to get the multiple inheritance to
  work across binaries.
- When exporting WorkQueue, it will export function with OSObjectPtr<dispatch_queue_t>.
  This has different type in Obj-C and in C++. If the type is used in both,
  then there will be unresolved symbol since C++ type is used when creating
  the library (WorkQueue.cpp), as such remove method prototype from WorkQueue
  and have a single WorkQueue constructor taking an enum instead.

Ideally, we would implement only bug 261080, but there are futher issues
getting it done. So to not block other works depending on this change,
we push this version first.

* Source/WTF/wtf/FunctionDispatcher.h:
* Source/WTF/wtf/SuspendableWorkQueue.cpp:
(WTF::SuspendableWorkQueue::stateString):
* Source/WTF/wtf/SuspendableWorkQueue.h:
* Source/WTF/wtf/WorkQueue.cpp:
(WTF::WorkQueue::main):
(WTF::WorkQueue::dispatch):
(WTF::WorkQueue::assertIsCurrent const):
(WTF::ConcurrentWorkQueue::dispatch):
* Source/WTF/wtf/WorkQueue.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY): Deleted.
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueue::WorkQueue):
(WTF::WorkQueue::constructMainWorkQueue): Deleted.
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueue::WorkQueue):
(WTF::WorkQueue::constructMainWorkQueue): Deleted.

Canonical link: https://commits.webkit.org/267623@main
